File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/core/usecases/deductions/GetAllDeductions.kt
package dev.auguste.agni_api.core.usecases.deductions

import dev.auguste.agni_api.core.adapters.repositories.IRepository
import dev.auguste.agni_api.core.adapters.dto.QueryFilter
import dev.auguste.agni_api.core.entities.Deduction
import dev.auguste.agni_api.core.usecases.ListOutput
import dev.auguste.agni_api.core.usecases.deductions.dto.GetDeductionOutput
import dev.auguste.agni_api.core.usecases.interfaces.IUseCase

class GetAllDeductions(private val deductionRepo: IRepository<Deduction>): IUseCase<QueryFilter, ListOutput<GetDeductionOutput>> {

    override fun execAsync(input: QueryFilter): ListOutput<GetDeductionOutput> {
        val deductions = deductionRepo.getAll(input)

        return ListOutput(
            items = deductions.items.map {
                GetDeductionOutput(
                    id = it.id,
                    title = it.title,
                    description = it.description,
                    base = it.base,
                    mode = it.mode
                )
            },
            total = deductions.total
        )
    }
}

File: apps/agni-server_archived/src/core/interactions/deduction/getAllDeductions.ts
import { ListDto, QueryFilter } from "@core/dto/base";
import { IUsecase } from "../interfaces";
import Repository from "@core/adapters/repository";
import { Deduction, DeductionType } from "@core/domains/entities/decution";

export type GetAllDeductionDto = {
    id: string
    typeId: string
    typeName: string
    rate: number
}

export class GetAllDeductionUseCase implements IUsecase<QueryFilter, ListDto<GetAllDeductionDto> > {
    private deductionTypeRepo: Repository<DeductionType>
    private deductionRepo: Repository<Deduction>

    constructor(deductionRepo: Repository<Deduction>, deductionTypeRepo: Repository<DeductionType>) {
        this.deductionTypeRepo = deductionTypeRepo 
        this.deductionRepo = deductionRepo
    }

    async execute(filter: QueryFilter, trx?: any): Promise<ListDto<GetAllDeductionDto>> {
        try {

            const deductions = await this.deductionRepo.getAll({ offset: filter.offset, limit: filter.limit }) 
            const deductionTypes = await this.deductionTypeRepo.getManyByIds(deductions.items.map(i => i.getDeductionTypeId()))

            return {
                items: deductions.items.map(i => ({
                    id: i.getId(), 
                    rate: i.getRate(),
                    typeId: i.getDeductionTypeId(),
                    typeName: deductionTypes.find(d => d.getId() === i.getDeductionTypeId())?.getTitle() || 'Unknow'
                })),
                totals: deductions.total
            }    
        } catch(err) {
            throw err
        }
    }
}